### PR TITLE
Fix commonjs exports with destructuring.

### PR DESCRIPTION
**Pull request by yavorsky**
| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          |  y
| Major: Breaking Change?  | n
| Minor: New Feature?      | n
| Deprecations?            | n
| Spec Compliancy?         | y
| Tests Added/Pass?        | y
| Fixed Tickets            | Fixes https://github.com/babel/babel-preset-env/issues/193
| License                  | MIT
| Doc PR                   |
| Dependency Changes       | 

[Original issue](https://github.com/babel/babel-preset-env/issues/193#issuecomment-284769720)
While we are exporting VariableDeclaration with ObjectPattern inside using just `transform-es2015-modules-commonjs` we are loosing ExportsAssignment:
*In*
```
export const { baz } = {};
```
*Out*
```
const { baz } = {};
```
*Expected*
```
const { baz } = {};
exports.baz = baz;
```

It works as expected with `transform-es2015-destructuring`:
*Out with transform-es2015-destructuring*
```
var _ref = {};
const baz = _ref.baz;
exports.baz = baz;
```

Also 1 test was added.

### CODE REVIEW

**Comment by aaronang** [comment]
Looks good to me except that the test is failing.

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 1]
```diff
@@ -306,6 +306,15 @@ export default function () {
                       addTo(exports, id.node.name, id.node);
                       init.replaceWith(buildExportsAssignment(id.node, init.node).expression);
                       nonHoistedExportNames[id.node.name] = true;
+                    } else if (id.isObjectPattern()) {
+                      for (let i = 0; i < id.node.properties.length; i++) {
+                        const prop = id.node.properties[i];
+                        if (!t.isRestProperty(prop)) {
```
Comment (danez) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Shouldn't the rest property also be handled?

for example
```
export const {foo: test, ...tester} = {};
```

Comment (yavorsky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: @danez Probably yes, but for now: 1) we haven't envs to use it without `transform-object-rest-spread` and 2) seems like we need `babel-plugin-syntax-object-rest-spread` for rest properties: [repl](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=env&targets=Node-7.4&browsers=&builtIns=false&code=export%20const%20%7Bfoo%3A%20test%2C%20...tester%7D%20%3D%20%7B%7D%3B&experimental=true&loose=true&spec=false&playground=false&stage=0)

Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: It's fine for people to have enabled support for the syntax without enabling the actual transformation, so I think it would be good to handle.

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 2]
```diff
@@ -306,6 +306,15 @@ export default function () {
                       addTo(exports, id.node.name, id.node);
                       init.replaceWith(buildExportsAssignment(id.node, init.node).expression);
                       nonHoistedExportNames[id.node.name] = true;
+                    } else if (id.isObjectPattern()) {
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Should we be handling `ArrayPattern` too?

Comment (yavorsky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: @loganfsmyth Sure! Going to add. Thanks!

Comment (xtuc) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Could the `ArrayPattern` be a separate PR?

Comment (yavorsky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: @xtuc I've started work on it but could pick it out to a separate PR

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 3]
```diff
@@ -129,6 +129,10 @@ export default function () {
   return {
     inherits: require("babel-plugin-transform-strict-mode"),
 
+    manipulateOptions(opts, parserOpts) {
+      parserOpts.plugins.push("objectRestSpread");
+    },
```
Comment (xtuc) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Would be better to inherits from `babel-plugin-syntax-object-rest-spread`. It seems multiple inheritance is not possible (at least didn't found an example).

To avoid that duplicated code, we could merge the config with [the syntax plugin](https://github.com/babel/babel/blob/8a82cc060ae0ab46bf52e05e592de770bd246f6f/packages/babel-plugin-syntax-object-rest-spread/src/index.js#L3-L5). What do you think?

Comment (yavorsky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: @xtuc Yes! Unfortunately, I didn't found multiple inheritance too. Can you please point me if there util for merging configs?

Comment (xtuc) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: It seems there are no merging util for configuration. I was thinking to use `Object.assign`.

Comment (yavorsky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Yes, just added Object.assign with `babel-plugin-transform-object-assign` as devDep to support node < 4.

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 4]
```diff
@@ -126,289 +129,287 @@ export default function () {
...
-
-        if (
-          state.opts.allowTopLevelThis !== true &&
-          !path.findParent((path) => !path.is("shadow") &&
-          THIS_BREAK_KEYS.indexOf(path.type) >= 0)
-        ) {
-          path.replaceWith(t.identifier("undefined"));
-        }
-      },
+  return Object.assign(
```
Comment (Kovensky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: This only works by coincidence because no keys clash, no actual merging is done.

Maybe it's ok in this case, but it's still... weird 😕

Comment (yavorsky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: yeah, I agree. Do you know other ways to inherit from multiple plugins?

Comment (Kovensky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Not really :cry:

Comment (yavorsky) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: I think it's worth to add this feature to core. Going to try today 😕

File: packages/babel-plugin-transform-es2015-modules-commonjs/test/fixtures/strict/export-const-destructuring-object-rest/actual.js [hunk 5]
```diff
@@ -1 +0,0 @@
-export const { foo, ...bar } = {};
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-commonjs/test/fixtures/strict/export-const-destructuring-object-rest/actual.js]: This test specifically seems like a reasonable thing to have, if you'd like to keep it as a devDep for the test.

**Comment by idris** [comment]
Ping! Looks like this is done. Can it be merged?

**Comment by mpyw** [comment]
I'm also waiting this PR merged...

**Comment by jridgewell** [comment]
This will also need to be ported to 7.0.

**Comment by yavorsky** [comment]
@jridgewell yes! on it.

**Comment by Yeti-or** [comment]
Guys, will it be released under `6.x` ?
